### PR TITLE
Role generation

### PR DESCRIPTION
**Pull request by ThomasKopsch**
I introduced different roles:
Now there is a new abstract class User from which the classes Customer and banker inherit. This will allow us to implement different functionality for different roles.

I suggest to implement a command SHOWACCOUNTS <CustomerID> which allows bankers to see the account of any customer.

### CODE REVIEW

**Review by vc409 — COMMENTED** [review]
Hi @ThomasKopsch ,

I have checked the code and I have the following comments:
- Please use CamelCase Java convention, instead of banker use Banker, etc.
- I think the method getUserType in the abstract class called User should be abstract so that you force the "child" to implement it.
- The class User defines a method setNewPassword where it compares the strings based on their JVM reference, if you wanted to compare the string content please use the **.equals** method of class String, and place the boolean expression in a single return, instead of an if-else, something like **return s1.equals(s2);**
- The class User can omit the constructor and the fields could be protected.

File: src/newbank/server/User.java [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+	private String userID;
+	private String password;
+	
+	public User(String userID, String password) {
+		this.userID = userID.toLowerCase();
+		this.password = password;
+	}
+	
+	public String getUserType() {
+		return "generic";
```
Comment (kelinj) [inline src/newbank/server/User.java]: We can have this as abstract method. 
	`abstract public String getUserType() ;`
and not let a "generic" user type exist. The type will be implemented by the child classes

File: src/newbank/server/banker.java [hunk 2]
```diff
@@ -0,0 +1,21 @@
+package newbank.server;
+
+import java.util.ArrayList;
+
+public class banker extends User {
```
Comment (kelinj) [inline src/newbank/server/banker.java]: Preferred to have Title case for Classes

File: src/newbank/server/userDB.java [hunk 3]
```diff
@@ -5,49 +5,49 @@
 
 import java.util.HashMap;
 
-public class customerDB {
+public class userDB {
```
Comment (kelinj) [inline src/newbank/server/userDB.java]: Nice work generalizing this. 
 Preferred to use Title case for class

File: src/newbank/server/NewBank.java [hunk 4]
```diff
@@ -4,14 +4,14 @@
...
 	
 	public NewBank() {
-		customers = new customerDB();
+		users = new userDB();
 	}
 	
-	public synchronized Customer checkLogInDetails(String customerID, String password) {
-		Customer customer = customers.getCustomer(customerID);
+	public synchronized User checkLogInDetails(String userID, String password) {
+		User customer = users.getUser(userID);
```
Comment (kelinj) [inline src/newbank/server/NewBank.java]: `User user= users.getUser(userID);` 
Since generic

File: src/newbank/server/NewBank.java [hunk 5]
```diff
@@ -22,15 +22,30 @@ public synchronized Customer checkLogInDetails(String customerID, String passwor
...
-				String accountName = request.substring( request.indexOf(" ")+1);
-				return makeNewAccount(customer, accountName);
+				if (user.getUserType().equals("customer")) {
+					Customer customer = (Customer) user;
+					String accountName = request.substring( request.indexOf(" ")+1);
+					return makeNewAccount(customer, accountName);				
+				}
+			//Execute SHOWACCOUNTS if the user is a customer			
+			} else if (request.startsWith("SHOWACCOUNTS")) {
+				if (user.getUserType().equals("banker")) {
```
Comment (kelinj) [inline src/newbank/server/NewBank.java:42]: Yeah true.  I can see a lot more like this which can be put in backlog specific to Banker 
Depositing money also could be something which can be Banker specific

File: src/newbank/server/User.java (hunk 1 above)
Comment (ThomasKopsch) [reply src/newbank/server/User.java]: Ok, I agree.

File: src/newbank/server/banker.java (hunk 2 above)
Comment (ThomasKopsch) [reply src/newbank/server/banker.java]: Agree.

File: src/newbank/server/userDB.java (hunk 3 above)
Comment (ThomasKopsch) [reply src/newbank/server/userDB.java]: Agree.

File: src/newbank/server/NewBank.java (hunk 4 above)
Comment (ThomasKopsch) [reply src/newbank/server/NewBank.java]: Agree.

**Pull request merged**